src/main/java/ev3dev/utils/DataChannelRereader.java

```diff
@@ -0,0 +1,71 @@
+package ev3dev.utils;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/**
+ * Reader of streams that can reread the same channel for structured data of
+ * known length. The focus of this class is on performance.
+ *
+ * @author David Walend
+ */
+public class DataChannelRereader implements Closeable {
+
+    private final Path path;
+    private final ByteBuffer byteBuffer;
+    private final FileChannel channel;
+
+    /**
+     * Create a DataChannelRereader for path with a bufferLength byte buffer
+     *
+     * @param path path to the file to reread
+     * @param bufferLength length of the buffer to hold the structure
+     * @throws IOException when things go wrong
+     */
+    public DataChannelRereader(Path path, int bufferLength) throws IOException {
+        this.path = path;
+        this.byteBuffer = ByteBuffer.allocate(bufferLength);
+        this.channel = FileChannel.open(path);
+    }
+
+    /**
+     * Create a DataChannelRereader for pathString with the default 32-byte buffer.
+     *
+     * @param pathString Path to the file to reread
+     * @throws IOException when things go wrong
+     */
+    public DataChannelRereader(String pathString) throws IOException {
+        this(Paths.get(pathString),32);
+    }
+
+    /**
+     * @return a string made from the bytes in the file;
+     */
+    public String readString() {
+        try {
+            int n;
+            do {
+                byteBuffer.clear();
+                channel.position(0);
+                n = channel.read(byteBuffer);
```

Comment:
I think this could be simplified to a `channel.read(byteBuffer, 0)` call. However, it's entirely possible that under the hood the two-arg read is just the following, which could be slower:
```
int old = channel.position();
channel.position(0);
int rval = channel.read(buffer);
channel.position(old);
return rval;
```